spring-integration-mqtt/src/test/java/org/springframework/integration/mqtt/ClientManagerBackToBackTests.java

```diff
@@ -102,6 +131,40 @@ private void testSubscribeAndPublish(Class<?> configClass, String topicName, Cou
 		}
 	}
 
+	private void testSubscribeAndPublishRuntime(Class<?> configClass, String topicName, CountDownLatch subscribedLatch, Class<?> adapter)
+			throws Exception {
+
+		try (var ctx = new AnnotationConfigApplicationContext(configClass)) {
+			// given
+			var input = ctx.getBean("mqttOutFlow.input", MessageChannel.class);
+			var flowContext = ctx.getBean(IntegrationFlowContext.class);
+			var clientManager = ctx.getBean(ClientManager.class);
+			var output = new QueueChannel();
+			Class<?>[] parameterTypes = {ClientManager.class, String[].class};
+			Constructor<?> declaredConstructor = adapter.getConstructor(parameterTypes);
+			flowContext.registration(IntegrationFlow
+					.from((MessageProducerSupport) declaredConstructor.newInstance(clientManager, new String[] {topicName}))
+					.channel(output)
+					.get()).register();
+			String testPayload = "foo";
+			assertThat(subscribedLatch.await(20, TimeUnit.SECONDS)).isTrue();
+
+			// when
+			input.send(MessageBuilder.withPayload(testPayload).setHeader(MqttHeaders.TOPIC, topicName).build());
+			Message<?> receive = output.receive(20_000);
+
+			// then
+			assertThat(receive).isNotNull();
+			Object payload = receive.getPayload();
+			if (payload instanceof String sp) {
+				assertThat(sp).isEqualTo(testPayload);
+			}
+			else {
+				assertThat(payload).isEqualTo(testPayload.getBytes(StandardCharsets.UTF_8));
```

Comment:
Why does this happen?
Why do we need this `if..else` logic?
Why the payload cannot always be string to simplify the things?

Reply:
To be honest I have just copied previously used tests but It is not possible, because `MqttPahoMessageDrivenChannelAdapter` returns `String`  and `Mqttv5PahoMessageDrivenChannelAdapter` returns `byte[]`. I think that it is due to changes between `MQTT v3` and `MQTT v5`, have not researched it.